### PR TITLE
bsunanda:Run2-alca39 Add the GammaJetSelector which is more suitable for the AlCaReco step…

### PR DESCRIPTION
**Pull request by bsunanda**
… for calibration using gamma+jet

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by mmusich** [comment]
Hello @bsunanda I don't understand the purpose of this PR. Can you clarify if this is intended as a new AlCaReco?

**Comment by bsunanda** [comment]
This was alternative to AlCaReco for Gamma+Jet as proposed by Andrius quite a while ago. It will replace eventually GammaJetAlCaRecoProducer. Andrius showed the results from this several months ago. I realized that it is not included in 8_0 or 8_1 releases and discussed with Andrius in putting this in. Once integrated, we will test and most likely change the AlCaReco step for HcalGammaJet.
________________________________

**Comment by bsunanda** [comment]
@mmusich Did you see my earlier explanation.

**Comment by mmusich** [comment]
+1

File: Calibration/HcalAlCaRecoProducers/plugins/AlCaGammaJetSelector.cc [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+  // Check the conditions for a good event
+  if (photons.size()<1) return false;
+  if (pfjets.size()<1) return false;
+
+  // Check the requirement for minimum pT
+  // Assume ordered collections, but make no assumption about the direction
+
+  if ((photons.begin()->pt() < minPtPhoton_) &&
+      (photons.back().pt() < minPtPhoton_)) {
```
Comment (davidlange6) [inline Calibration/HcalAlCaRecoProducers/plugins/AlCaGammaJetSelector.cc]: @bsunanda - I don't understand this logic?

File: Calibration/HcalAlCaRecoProducers/plugins/AlCaGammaJetSelector.cc [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    int found=0;
+    for (reco::PhotonCollection::const_iterator it= photons.begin()+1;
```
Comment (davidlange6) [inline Calibration/HcalAlCaRecoProducers/plugins/AlCaGammaJetSelector.cc]: @bsunanda - assumes 2 photons in the list - this is not in general the case.

**Comment by bsunanda** [comment]
@davidlange6 I initially wrote a simpler selector function assuming no ordering - but Andrius wrote the new selector and tested with MC. The simpler one was in HcalAlCaGammaJetProducer.cc in the src directory. If you want I can substitute that one here. Both select the same event set. The other was something like
  // Check the requirement for minimum pT
  if (ph.size() == 0) return false;
  bool ok(false);
  for (reco::PFJetCollection::const_iterator itr=jt.begin();
       itr!=jt.end(); ++itr) {
    if (itr->pt() >= minPtJet_) {
      ok = true;
      break;
    }
  }
  if (!ok) return ok;
  for (reco::PhotonCollection::const_iterator itr=ph.begin();
       itr!=ph.end(); ++itr) {
    if (itr->pt() >= minPtPhoton_) return ok;
  }
  return false;

**Comment by davidlange6** [comment]
@bsunanda - I don't think the two code snippets are the same. (I understand the one in your comment.. not the one in your PR)

**Comment by bsunanda** [comment]
@davidlange66 So I shall change the code to the one in the comment
________________________________

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@mmusich Please approve the modified version (After David's comments)

**Comment by mmusich** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**